### PR TITLE
run Msbuild.Integration.Tests in parallel

### PR DESCRIPTION
**Pull request by kartheekp-ms**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/1285

Regression? Last working version:

## Description

Added `UseParallelXunit` property to `Msbuild.Integration.Tests` project so that the tests can run in parallel using the settings defined in the shared [`xunit.runner.json`](https://github.com/NuGet/NuGet.Client/blob/dev/build/TestShared/xunit.runner.json) file. `Msbuild.Integration.Tests` project has `28` tests out of which `27` tests are in a single class file and the remaining one in a separate class. Given that `xUnit` runs tests in a single class sequentially, I have moved `10` tests related to `package source mapping feature` in a new class file to get the benefit of parallelism. I have also removed existing `CollectionDefinition` because xUnit runs all the tests classes decorated the same collection sequentially. Added `IClassFixture` to share context for every test class.

From the CI logs below, `xUnit` started `3 tests in parallel (one test from every class file).

``` 
Discovering: Msbuild.Integration.Test (app domain = on [shadow copy], method display = ClassAndMethod, method display options = None)
           Discovered:  Msbuild.Integration.Test (found 28 test cases)
           Starting:    Msbuild.Integration.Test (parallel test collections = on, max threads = 8)
             Msbuild.Integration.Test.GenerateRestoreGraphFileTargetTests.GenerateRestoreGraphFile_WithMixedProjectSolution_BothStaticGraphAndStandard_AreEqual [STARTING]
             Msbuild.Integration.Test.MsbuildRestoreTaskTests.MsbuildRestore_WithCPPCliVcxproj_WithNativeAndManagedTransitiveDependency_Succeeds [STARTING]
             Msbuild.Integration.Test.MsbuildRestoreTaskPackageSourceMappingTests.MsbuildRestore_PackageSourceMappingFull_Succeed [STARTING]

```

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] N/A 

- **Documentation**
  
  - [x] N/A

### CODE REVIEW

**Comment by erdembayar** [comment]
Same comment as [other one.](https://github.com/NuGet/NuGet.Client/pull/4398#issuecomment-1009243240)

**Review by jeffkl — APPROVED** [review]

File: test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskPackageSourceMappingTests.cs [hunk 1]
```diff
@@ -0,0 +1,988 @@
...
+using System.Threading.Tasks;
+using NuGet.Frameworks;
+using NuGet.Packaging;
+using NuGet.ProjectModel;
+using NuGet.Test.Utility;
+using Xunit;
+
+namespace Msbuild.Integration.Test
+{
+    public class MsbuildRestoreTaskPackageSourceMappingTests : IClassFixture<MsbuildIntegrationTestFixture>
```
Comment (zivkan) [inline test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskPackageSourceMappingTests.cs:16]: Why is this file in the diff?  It doesn't appear to be a rename from another file, and adding PackageSourceMapping tests is out of scope for this PR (making tests run in parallel)

Comment (zivkan) [reply test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskPackageSourceMappingTests.cs:16]: Never mind, I didn't read the PR comment carefully enough, or the othe test file diff.

**Review by zivkan — APPROVED** [review]

**Comment by kartheekp-ms** [comment]
I triggered 4 CI builds to run only required stages. 3 out of 4 builds passed but 1 build failed `warning MSB4181: The "Exec" task returned false but did not log an error. `  Looking at the failure logs, Msbuild integration tests ran in parallel, and all the tests passed. The job failed with this error later.

image

This CI build for this pull request also failed with the same error but looking at the error logs all the test projects successfully completed execution.

```
=== TEST EXECUTION SUMMARY ===
            Msbuild.Integration.Test             Total:   31, Errors: 0, Failed: 0, Skipped:  0, Time:   40.585s
            NuGet.CommandLine.FuncTest           Total:  361, Errors: 0, Failed: 0, Skipped:  0, Time:  712.896s
            NuGet.CommandLine.Test               Total: 1033, Errors: 0, Failed: 0, Skipped: 39, Time:  759.234s
            NuGet.Commands.FuncTest              Total:   89, Errors: 0, Failed: 0, Skipped:  2, Time:  109.517s
            NuGet.MSSigning.Extensions.FuncTest  Total:   20, Errors: 0, Failed: 0, Skipped:  0, Time:   12.776s
            NuGet.PackageManagement.Test         Total:  445, Errors: 0, Failed: 0, Skipped:  4, Time:  157.507s
            NuGet.Packaging.FuncTest             Total:  228, Errors: 0, Failed: 0, Skipped:  0, Time:  382.445s
            NuGet.Protocol.FuncTest              Total:   51, Errors: 0, Failed: 0, Skipped: 16, Time:   52.164s
            NuGet.XPlat.FuncTest                 Total:  336, Errors: 0, Failed: 0, Skipped:  3, Time:   30.344s
                                                        ----          -          -           --        ---------
                                           GRAND TOTAL: 2594          0          0           64        2257.470s (2266.876s)
##[warning]build\build.proj(420,5): Warning MSB3073: The command  
     1>s\build\build.proj(420,5): warning MSB4181: The "Exec" task returned false but did not log an error.
         The previous error was converted to a warning because the task was called with ContinueOnError=true.
         Build continuing because "ContinueOnError" on the task "Exec" is set to "true".
##[error]build\build.proj(437,5): Error : Desktop CoreFuncTests tests failed!
```
@jeffkl - Is this failure somehow related to https://github.com/dotnet/msbuild/issues/5508?

**Comment by jeffkl** [comment]
> Is this failure somehow related to dotnet/msbuild#5508?

Sounds like a good assumption, is it one particular test that had this message?

**Comment by kartheekp-ms** [comment]
> Sounds like a good assumption, is it one particular test that had this message?

No.The "Exec" task returned this warning with no additional test message. I assume if it is a test failure then AzDO will report it under `Tests` tab for the build which was not the case. 100% tests passed.

image

**Comment by erdembayar** [comment]
> > Is this failure somehow related to [dotnet/msbuild#5508](https://github.com/dotnet/msbuild/issues/5508)?
> 
> Sounds like a good assumption, is it one particular test that had this message?

Just in case could you queue few more today? Yesterday many CI build failed, hard to see what was actually causing.

**Comment by kartheekp-ms** [comment]
Triggered 5 (1 automatic and 4 manual) builds on CI and all the required phases passed without any retries.

image

**Comment by erdembayar** [comment]
Main intention for turning on "parallel" is gain perf speed. Does this decrease CI/CD build duration?

**Comment by kartheekp-ms** [comment]
| Job Name | Before | After |
| ------------- | ------------- |------------- |
| Functional_Tests_On_Windows IsDektop  | 53.353 seconds (last official build - 6.2.0.10) | 41.149 seconds (private build - 6.2.0.42) |

Overall, this fix reduced `MsBuild.Integration.FuncTest` project test execution time by `12.204 seconds`. These tests are only configured to run for `net472` framework. Hence only one job is listed in the above table.

**Review by erdembayar — APPROVED** [review]

**Pull request merged**